test/framework/autoscaler_helpers.go

```diff
@@ -0,0 +1,426 @@
+type AddScaleUpDeploymentAndWaitInput struct {
+	ClusterProxy ClusterProxy
+}
+
+// AddScaleUpDeploymentAndWait create a deployment that will trigger the autoscaler to scale up and create a new machine.
+func AddScaleUpDeploymentAndWait(ctx context.Context, input AddScaleUpDeploymentAndWaitInput, intervals ...interface{}) {
+	By("Create a scale up deployment with resource requests to force scale up")
+
+	// gets the node size
+	nodes := &corev1.NodeList{}
+	workers := 0
+	Expect(input.ClusterProxy.GetClient().List(ctx, nodes)).To(Succeed(), "failed to list nodes")
+	var memory *resource.Quantity
+	for _, n := range nodes.Items {
+		if _, ok := n.Labels[nodeRoleControlPlane]; ok {
+			continue
+		}
+		if _, ok := n.Labels[nodeRoleOldControlPlane]; ok {
+			continue
+		}
+		memory = n.Status.Capacity.Memory() // Assume that all nodes have the same memory.
+		workers++
+	}
+	Expect(memory).ToNot(BeNil(), "failed to get memory for the worker node")
+
+	// creates a deployment requesting more memory than the worker has, thus triggering autoscaling
+	// Each pod should requests memory resource of about 60% of the node capacity so that at most one pod
+	// fits on each node. Setting a replicas of workers + 1 would ensure we have pods that cannot be scheduled.
+	// This will force exactly one extra node to be spun up.
+	replicas := workers + 1
+	memoryRequired := int64(float64(memory.Value()) * 0.6)
+	podMemory := resource.NewQuantity(memoryRequired, resource.BinarySI)
+
+	scalelUpDeployment := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "scale-up",
+			Namespace: metav1.NamespaceDefault,
+			Labels: map[string]string{
+				"app": "scale-up",
+			},
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas: pointer.Int32(int32(replicas)),
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"app": "scale-up",
+				},
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"app": "scale-up",
+					},
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:  "busybox",
+							Image: "busybox",
+							Resources: corev1.ResourceRequirements{
+								Requests: map[corev1.ResourceName]resource.Quantity{
+									corev1.ResourceMemory: *podMemory,
+								},
+							},
+							Command: []string{"/bin/sh", "-c", "echo \"up\" & sleep infinity"},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	By("Scale up deployment created")
+	Expect(input.ClusterProxy.GetClient().Create(ctx, scalelUpDeployment)).To(Succeed(), "failed to create the scale up pod")
+
+	By("Wait for the scale up deployment to become ready (this implies machines to be created)")
+	WaitForDeploymentsAvailable(ctx, WaitForDeploymentsAvailableInput{
+		Getter:     input.ClusterProxy.GetClient(),
+		Deployment: scalelUpDeployment,
+	}, intervals...)
+}
+
+type ProcessYAMLInput struct {
```

Comment:
I'm fine with this for now, but can we please open a follow-up issue to replace this with simple go templating?
(we already do it in other parts of our test framework (debian injection script)).

I think this would be a lot simpler than using clusterctl for it

Reply:
Follow-up issue: https://github.com/kubernetes-sigs/cluster-api/issues/8592